AGP Upgrade Assistant • Build Failure

Hello @jiusanzhou,

On the latest version of the package I'm getting the following **error**:

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':flutter_notification_listener'.
> Could not create an instance of type com.android.build.api.variant.impl.LibraryVariantBuilderImpl.
   > Namespace not specified. Specify a namespace in the module's build file. See https://d.android.com/r/tools/upgrade-assistant/set-namespace for information about setting the namespace.

     If you've specified the package attribute in the source AndroidManifest.xml, you can use the AGP Upgrade Assistant to migrate to the namespace value in the build file. Refer to https://d.android.com/r/tools/upgrade-assistant/agp-upgrade-assistant for general information about using the AGP Upgrade Assistant.
```

I also tried to fork the repo and fix the issue using the official [documentation provided by the Android Studio guidelines](https://developer.android.com/build/agp-upgrade-assistant) but without success.

Thanks a lot for your time ✨

--- a/README.md
+++ b/README.md
@@ -46,7 +46,11 @@ The plugin uses an Android system service to track notifications. To allow this 
 ```xml
 <service android:name="im.zoe.labs.flutter_notification_listener.NotificationsHandlerService"
     android:label="Flutter Notifications Handler"
-    android:permission="android.permission.BIND_NOTIFICATION_LISTENER_SERVICE">
+    android:permission="android.permission.BIND_NOTIFICATION_LISTENER_SERVICE"
+    android:exported="true"
+    android:foregroundServiceType="specialUse">
+    <property android:name="android.app.PROPERTY_SPECIAL_USE_FGS_SUBTYPE"
+        android:value="notification_listener"/>
     <intent-filter>
         <action android:name="android.service.notification.NotificationListenerService" />
     </intent-filter>
@@ -57,7 +61,13 @@ And don't forget to add the permissions to the manifest,
 ```xml
 <uses-permission android:name="android.permission.WAKE_LOCK" />
 <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
+<!-- Required for Android 13+ (API 33) to show foreground service notification -->
+<uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
+<!-- Required for Android 14+ (API 34) for foreground services -->
+<uses-permission android:name="android.permission.FOREGROUND_SERVICE_SPECIAL_USE"/>
 ```
+
+> **Note for Android 13+**: You need to request the `POST_NOTIFICATIONS` permission at runtime before starting the foreground service. See the example app for implementation details.
 
 **2. Init the plugin and add listen handler**
 
@@ -112,7 +122,8 @@ It's every useful while you want to start listening notifications automatically 
 Register a broadcast receiver in the `AndroidManifest.xml`,
 ```xml
 <receiver android:name="im.zoe.labs.flutter_notification_listener.RebootBroadcastReceiver"
-    android:enabled="true">
+    android:enabled="true"
+    android:exported="true">
     <intent-filter>
         <action android:name="android.intent.action.BOOT_COMPLETED" />
     </intent-filter>
@@ -126,7 +137,11 @@ And don't forget to add the permissions to the manifest,
 ```xml
 <uses-permission android:name="android.permission.WAKE_LOCK" />
 <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
-<!-- this pemission is for auto start service after reboot -->
+<!-- Required for Android 13+ (API 33) to show foreground service notification -->
+<uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
+<!-- Required for Android 14+ (API 34) for foreground services -->
+<uses-permission android:name="android.permission.FOREGROUND_SERVICE_SPECIAL_USE"/>
+<!-- This permission is for auto start service after reboot -->
 <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
 ```
 

--- a/android/build.gradle
+++ b/android/build.gradle
@@ -2,14 +2,14 @@ group 'im.zoe.labs.flutter_notification_listener'
 version '1.0-SNAPSHOT'
 
 buildscript {
-    ext.kotlin_version = '1.9.0'
+    ext.kotlin_version = '2.1.0'
     repositories {
         google()
-        jcenter()
+        mavenCentral()
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.3.0'
+        classpath 'com.android.tools.build:gradle:8.7.0'
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
     }
 }
@@ -17,7 +17,7 @@ buildscript {
 rootProject.allprojects {
     repositories {
         google()
-        jcenter()
+        mavenCentral()
     }
 }
 
@@ -25,16 +25,26 @@ apply plugin: 'com.android.library'
 apply plugin: 'kotlin-android'
 
 android {
-    compileSdkVersion 34
+    namespace 'im.zoe.labs.flutter_notification_listener'
+    compileSdk 34
 
     sourceSets {
         main.java.srcDirs += 'src/main/kotlin'
     }
     defaultConfig {
-        minSdkVersion 23
+        minSdk 21
+    }
+
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_17
+        targetCompatibility JavaVersion.VERSION_17
+    }
+
+    kotlinOptions {
+        jvmTarget = '17'
     }
 }
 
 dependencies {
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
 }

--- a/android/gradle/wrapper/gradle-wrapper.properties
+++ b/android/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,5 @@
-#Sat Jun 05 09:48:27 CST 2021
 distributionBase=GRADLE_USER_HOME
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.1-all.zip
 distributionPath=wrapper/dists
-zipStorePath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.9-all.zip

--- a/android/src/main/AndroidManifest.xml
+++ b/android/src/main/AndroidManifest.xml
@@ -1,7 +1,10 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:tools="http://schemas.android.com/tools"
-    package="im.zoe.labs.flutter_notification_listener">
+    xmlns:tools="http://schemas.android.com/tools">
     <uses-permission android:name="android.permission.WAKE_LOCK" />
     <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
+    <!-- Required for Android 13+ (API 33) to show foreground service notification -->
+    <uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
+    <!-- Required for Android 14+ (API 34) for foreground services -->
+    <uses-permission android:name="android.permission.FOREGROUND_SERVICE_SPECIAL_USE"/>
 </manifest>

--- a/android/src/main/kotlin/im/zoe/labs/flutter_notification_listener/FlutterNotificationListenerPlugin.kt
+++ b/android/src/main/kotlin/im/zoe/labs/flutter_notification_listener/FlutterNotificationListenerPlugin.kt
@@ -60,9 +60,10 @@ class FlutterNotificationListenerPlugin : FlutterPlugin, MethodChannel.MethodCal
     intentFilter.addAction(NotificationsHandlerService.NOTIFICATION_INTENT)
     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
       mContext.registerReceiver(receiver, intentFilter, RECEIVER_EXPORTED)
-    }else {
+    } else {
       mContext.registerReceiver(receiver, intentFilter)
     }
+
     Log.i(TAG, "attached engine finished")
   }
 
@@ -73,7 +74,7 @@ class FlutterNotificationListenerPlugin : FlutterPlugin, MethodChannel.MethodCal
       methodChannel = null 
     }
 
-    val event = eventChannel
+    val even= eventChannel
     if (event != null) {
       event.setStreamHandler(null) 
       eventChannel = null 

--- a/android/src/main/kotlin/im/zoe/labs/flutter_notification_listener/FlutterNotificationListenerPlugin.kt
+++ b/android/src/main/kotlin/im/zoe/labs/flutter_notification_listener/FlutterNotificationListenerPlugin.kt
@@ -74,7 +74,7 @@ class FlutterNotificationListenerPlugin : FlutterPlugin, MethodChannel.MethodCal
       methodChannel = null 
     }
 
-    val even= eventChannel
+    val event = eventChannel
     if (event != null) {
       event.setStreamHandler(null) 
       eventChannel = null 

--- a/android/src/main/kotlin/im/zoe/labs/flutter_notification_listener/NotificationEvent.kt
+++ b/android/src/main/kotlin/im/zoe/labs/flutter_notification_listener/NotificationEvent.kt
@@ -138,7 +138,8 @@ class NotificationEvent(context: Context, sbn: StatusBarNotification) {
 
         @RequiresApi(Build.VERSION_CODES.M)
         private fun convertIconToByteArray(context: Context, icon: Icon): ByteArray {
-            return convertBitmapToByteArray(icon.loadDrawable(context)!!.toBitmap())
+            val drawable = icon.loadDrawable(context) ?: return ByteArray(0)
+            return convertBitmapToByteArray(drawable.toBitmap())
         }
 
         private fun convertBitmapToByteArray(bitmap: Bitmap): ByteArray {

--- a/example/android/app/build.gradle
+++ b/example/android/app/build.gradle
@@ -1,14 +1,15 @@
+plugins {
+    id "com.android.application"
+    id "kotlin-android"
+    id "dev.flutter.flutter-gradle-plugin"
+}
+
 def localProperties = new Properties()
 def localPropertiesFile = rootProject.file('local.properties')
 if (localPropertiesFile.exists()) {
     localPropertiesFile.withReader('UTF-8') { reader ->
         localProperties.load(reader)
     }
-}
-
-def flutterRoot = localProperties.getProperty('flutter.sdk')
-if (flutterRoot == null) {
-    throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
 }
 
 def flutterVersionCode = localProperties.getProperty('flutter.versionCode')
@@ -21,30 +22,33 @@ if (flutterVersionName == null) {
     flutterVersionName = '1.0'
 }
 
-apply plugin: 'com.android.application'
-apply plugin: 'kotlin-android'
-apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
-
 android {
-    compileSdkVersion 31
+    namespace 'im.zoe.labs.flutter_notification_listener_example'
+    compileSdk 34
 
     sourceSets {
         main.java.srcDirs += 'src/main/kotlin'
     }
 
     defaultConfig {
-        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
         applicationId "im.zoe.labs.flutter_notification_listener_example"
-        minSdkVersion 19
-        targetSdkVersion 30
+        minSdkVersion flutter.minSdkVersion
+        targetSdk 34
         versionCode flutterVersionCode.toInteger()
         versionName flutterVersionName
     }
 
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_17
+        targetCompatibility JavaVersion.VERSION_17
+    }
+
+    kotlinOptions {
+        jvmTarget = '17'
+    }
+
     buildTypes {
         release {
-            // TODO: Add your own signing config for the release build.
-            // Signing with the debug keys for now, so `flutter run --release` works.
             signingConfig signingConfigs.debug
         }
     }
@@ -52,8 +56,4 @@ android {
 
 flutter {
     source '../..'
-}
-
-dependencies {
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
 }

--- a/example/android/app/src/main/AndroidManifest.xml
+++ b/example/android/app/src/main/AndroidManifest.xml
@@ -1,9 +1,12 @@
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="im.zoe.labs.flutter_notification_listener_example">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
 
     <uses-permission android:name="android.permission.WAKE_LOCK" />
     <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
+    <!-- Required for Android 13+ (API 33) to show foreground service notification -->
+    <uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
+    <!-- Required for Android 14+ (API 34) for foreground services -->
+    <uses-permission android:name="android.permission.FOREGROUND_SERVICE_SPECIAL_USE"/>
 
    <application
         android:label="flutter_notification_listener_example"
@@ -11,6 +14,7 @@
        android:allowBackup="true">
         <activity
             android:name=".MainActivity"
+            android:exported="true"
             android:launchMode="singleTop"
             android:theme="@style/LaunchTheme"
             android:configChanges="orientation|keyboardHidden|keyboard|screenSize|smallestScreenSize|locale|layoutDirection|fontScale|screenLayout|density|uiMode"
@@ -54,7 +58,11 @@
        </receiver>
         <service android:name="im.zoe.labs.flutter_notification_listener.NotificationsHandlerService"
             android:label="Flutter Notifications Handler"
-            android:permission="android.permission.BIND_NOTIFICATION_LISTENER_SERVICE">
+            android:permission="android.permission.BIND_NOTIFICATION_LISTENER_SERVICE"
+            android:exported="true"
+            android:foregroundServiceType="specialUse">
+            <property android:name="android.app.PROPERTY_SPECIAL_USE_FGS_SUBTYPE"
+                android:value="notification_listener"/>
             <intent-filter>
                <action android:name="android.service.notification.NotificationListenerService" />
             </intent-filter>

--- a/example/android/build.gradle
+++ b/example/android/build.gradle
@@ -1,21 +1,7 @@
-buildscript {
-    ext.kotlin_version = '1.6.20'
-    // ext.kotlin_version = '1.3.50'
-    repositories {
-        google()
-        jcenter()
-    }
-
-    dependencies {
-        classpath 'com.android.tools.build:gradle:4.1.0'
-        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
-    }
-}
-
 allprojects {
     repositories {
         google()
-        jcenter()
+        mavenCentral()
     }
 }
 
@@ -27,6 +13,6 @@ subprojects {
     project.evaluationDependsOn(':app')
 }
 
-task clean(type: Delete) {
-    delete rootProject.buildDir
+tasks.register("clean", Delete) {
+    delete rootProject.layout.buildDirectory
 }

--- a/example/android/gradle.properties
+++ b/example/android/gradle.properties
@@ -1,3 +1,3 @@
-org.gradle.jvmargs=-Xmx1536M
+org.gradle.jvmargs=-Xmx4096M
 android.useAndroidX=true
-android.enableJetifier=true
+android.enableJetifier=false

--- a/example/android/gradle/wrapper/gradle-wrapper.properties
+++ b/example/android/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,5 @@
-#Fri Jun 23 08:50:38 CEST 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.7-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.9-all.zip

--- a/example/android/settings.gradle
+++ b/example/android/settings.gradle
@@ -1,11 +1,25 @@
-include ':app'
+pluginManagement {
+    def flutterSdkPath = {
+        def properties = new Properties()
+        file("local.properties").withInputStream { properties.load(it) }
+        def flutterSdkPath = properties.getProperty("flutter.sdk")
+        assert flutterSdkPath != null, "flutter.sdk not set in local.properties"
+        return flutterSdkPath
+    }()
 
-def localPropertiesFile = new File(rootProject.projectDir, "local.properties")
-def properties = new Properties()
+    includeBuild("$flutterSdkPath/packages/flutter_tools/gradle")
 
-assert localPropertiesFile.exists()
-localPropertiesFile.withReader("UTF-8") { reader -> properties.load(reader) }
+    repositories {
+        google()
+        mavenCentral()
+        gradlePluginPortal()
+    }
+}
 
-def flutterSdkPath = properties.getProperty("flutter.sdk")
-assert flutterSdkPath != null, "flutter.sdk not set in local.properties"
-apply from: "$flutterSdkPath/packages/flutter_tools/gradle/app_plugin_loader.gradle"
+plugins {
+    id "dev.flutter.flutter-plugin-loader" version "1.0.0"
+    id "com.android.application" version "8.7.0" apply false
+    id "org.jetbrains.kotlin.android" version "2.1.0" apply false
+}
+
+include ":app"

--- a/example/lib/main.dart
+++ b/example/lib/main.dart
@@ -34,6 +34,7 @@ class NotificationsLog extends StatefulWidget {
   _NotificationsLogState createState() => _NotificationsLogState();
 }
 
+@pragma('vm:entry-point')
 class _NotificationsLogState extends State<NotificationsLog> {
   List<NotificationEvent> _log = [];
   bool started = false;
@@ -48,7 +49,8 @@ class _NotificationsLogState extends State<NotificationsLog> {
   }
 
   // we must use static method, to handle in background
-  @pragma('vm:entry-point') // prevent dart from stripping out this function on release build in Flutter 3.x
+  @pragma(
+      'vm:entry-point') // prevent dart from stripping out this function on release build in Flutter 3.x
   static void _callback(NotificationEvent evt) {
     print("send evt to ui: $evt");
     final SendPort? send = IsolateNameServer.lookupPortByName("_listener_");

--- a/lib/src/plugin.dart
+++ b/lib/src/plugin.dart
@@ -164,6 +164,7 @@ class NotificationsListener {
 }
 
 /// callbackDispatcher use to install background channel
+@pragma('vm:entry-point')
 void callbackDispatcher({inited = true}) {
   WidgetsFlutterBinding.ensureInitialized();
 

